go/proxyd/cache.go

```diff
@@ -0,0 +1,275 @@
+		"eth_getBlockRange":    true,
+	}
+)
+
+var (
+	errInvalidBlockByNumberParams = errors.New("invalid eth_getBlockByNumber params")
+	errUnavailableBlockNumSyncer  = errors.New("getLatestBlockFn not set for required RPC")
+)
+
+type cache struct {
+	lru *lru.Cache
+}
+
+func newMemoryCache() *cache {
+	rep, _ := lru.New(memoryCacheLimit)
+	return &cache{rep}
+}
+
+func (c *cache) Get(ctx context.Context, key string) (string, error) {
+	if val, ok := c.lru.Get(key); ok {
+		return val.(string), nil
+	}
+	return "", nil
+}
+
+func (c *cache) Put(ctx context.Context, key string, value string) error {
+	c.lru.Add(key, value)
+	return nil
+}
+
+type redisCache struct {
+	rdb *redis.Client
+}
+
+func newRedisCache(url string) (*redisCache, error) {
+	opts, err := redis.ParseURL(url)
+	if err != nil {
+		return nil, err
+	}
+	rdb := redis.NewClient(opts)
+	if err := rdb.Ping(context.Background()).Err(); err != nil {
+		return nil, wrapErr(err, "error connecting to redis")
+	}
+	return &redisCache{rdb}, nil
+}
+
+func (c *redisCache) Get(ctx context.Context, key string) (string, error) {
+	val, err := c.rdb.Get(ctx, key).Result()
+	if err == redis.Nil {
+		return "", nil
+	} else if err != nil {
+		return "", err
+	}
+	return val, nil
+}
+
+func (c *redisCache) Put(ctx context.Context, key string, value string) error {
+	err := c.rdb.Set(ctx, key, value, 0).Err()
+	return err
+}
+
+type GetLatestBlockNumFn func(ctx context.Context) (uint64, error)
+
+type RPCCache struct {
+	cache               Cache
+	getLatestBlockNumFn GetLatestBlockNumFn
+}
+
+func newRPCCache(cache Cache, getLatestBlockNumFn GetLatestBlockNumFn) *RPCCache {
+	return &RPCCache{cache: cache, getLatestBlockNumFn: getLatestBlockNumFn}
+}
+
+func (c *RPCCache) GetRPC(ctx context.Context, req *RPCReq) (*RPCRes, error) {
+	if !c.isCacheable(req) {
+		return nil, nil
+	}
+
+	key := mustMarshalJSON(req)
+	encodedVal, err := c.cache.Get(ctx, string(key))
+	if err != nil {
+		return nil, err
+	}
+	if encodedVal == "" {
+		return nil, nil
+	}
+
+	val, err := snappy.Decode(nil, []byte(encodedVal))
+	if err != nil {
+		return nil, err
+	}
+
+	res := new(RPCRes)
+	err = json.Unmarshal(val, res)
+	if err != nil {
+		panic(err)
+	}
+
+	return res, nil
+}
+
+func (c *RPCCache) PutRPC(ctx context.Context, req *RPCReq, res *RPCRes) error {
+	if !c.isCacheable(req) {
+		return nil
+	}
+	if supportedBlockRPCMethods[req.Method] {
+		if ok, err := c.isConfirmed(ctx, req); err != nil {
+			return err
+		} else if !ok {
+			return nil
+		}
+	}
+
+	key := mustMarshalJSON(req)
+	val := mustMarshalJSON(res)
+	encodedVal := snappy.Encode(nil, val)
+	return c.cache.Put(ctx, string(key), string(encodedVal))
+}
+
+func (c *RPCCache) isConfirmed(ctx context.Context, req *RPCReq) (bool, error) {
+	if c.getLatestBlockNumFn == nil {
+		return false, errUnavailableBlockNumSyncer
+	}
+	curBlock, err := c.getLatestBlockNumFn(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	switch req.Method {
```

Comment:
Rather than using a big `switch` like this, I suggest we use a dispatch table that maps RPC methods to handlers:

1. Define an interface called `MethodHandler` or something with the following methods:

- `IsCacheable`
- `IsConfirmed`
- `CacheKey`

2. Define handlers for each method, then map each method to a `MethodHandler`.
3. Then, in the `GetRPC` / `PutRPC` methods, look up the `MethodHandler` via the dispatch table.

This will make it easier for us to define additional handlers in the future.